extensions/amp-addthis/0.1/ConfigManager.js

```diff
@@ -0,0 +1,244 @@
+
+    /**
+     * Flag; only a single request should register as a page view.
+     * @type {boolean}
+     * @private
+     */
+    this.registerView_ = true;
+  }
+
+  /**
+   * Store the configuration received for the provided pubId, mark the request for the config as
+   * completed, and broadcast the config to relevant iframes.
+   * @private
+   */
+  receiveConfiguration_({config, pubId}) {
+    const pubData = this.dataForPubId_[pubId];
+    pubData.config = config;
+    pubData.requestStatus = RequestStatus.COMPLETED;
+    const {iframes} = pubData;
+    iframes.forEach(iframe => this.sendConfiguration_({iframe, pubId}));
+  }
+
+  /** @private */
+  sendConfiguration_({iframe, pubId}) {
+    const {location: loc, title} = this.ownerDocument_;
+    const pubData = this.dataForPubId_[pubId];
+    const dashboardConfig = pubData.config;
+    const configRequestStatus = pubData.requestStatus;
+    const widgetId = iframe[WIDGET_ID_PROP];
+    const jsonToSend = /** @type JsonObject */ ({
+      event: CONFIGURATION_EVENT,
+      shareConfig: {
+        url: loc.href,
+        title,
+      },
+      pubId,
+      widgetId,
+      configRequestStatus,
+      dashboardConfig,
+      registerView: this.registerView_,
+    });
+
+    iframe.contentWindow./*OK*/postMessage(JSON.stringify(jsonToSend), ORIGIN);
+
+    if (this.registerView_) {
+      this.registerView_ = false;
+    }
+
+    if (configRequestStatus === RequestStatus.NOT_REQUESTED) {
+      // If a config for this pubId has not been requested yet, then this iframe will be the one
+      // responsible for requesting it and sending it back here.
+      this.configProviderIframes_.push(iframe);
+      pubData.requestStatus = RequestStatus.REQUESTED;
+    }
+  }
+
+  /**
+   * Handles messages posted from amp-addthis iframes, ensuring the correct origin, and ensuring
+   * that the iframe is one that is expected to provide configuration information.
+   * @private
+   */
+  handleAddThisMessage_(event) {
+    if (event.origin !== ORIGIN || !getData(event)) {
+      return;
+    }
+
+    const isProviderIframe = this.configProviderIframes_.some(iframe => {
+      return iframe.contentWindow === event.source;
+    });
+
+    if (!isProviderIframe) {
+      return;
+    }
+
+    const data = getMessageData(event) || {};
+
+    if (data.event === CONFIGURATION_EVENT) {
+      this.receiveConfiguration_(data);
+    }
+  }
+
+  /**
+   * Register relevant data with the configuration manager and prepare request/response cycle
+   * between frames.
+   * @param {{pubId:!string, iframe:!Element, iframeLoadPromise:!Promise, win:!EventTarget, element:!Element}} param
+   */
+  register({pubId, iframe, iframeLoadPromise, win, element}) {
+    if (!this.removeMessageListener_) {
+      this.removeMessageListener_ = listen(
+          win, 'message', this.handleAddThisMessage_.bind(this)
+      );
+    }
+
+    if (!this.ownerDocument_) {
+      this.ownerDocument_ = element.ownerDocument;
```

Comment:
This will have issues for embedding PWA, etc.

Reply:
@jridgewell Do you think you could say more about the issue here? For what it's worth, `element` in this case is just the AmpElement's `this.element`, and the intent is to get the title (etc.) set on the AmpElement's document using it. I could just pass each particular value that I'm interested in, instead, if the concern is an additional reference to the element's `ownerDocument`. Or is the concern that these values wouldn't always be accurate when the element is deeply embedded (or won't necessarily reflect the app's name, etc., in the case of a PWA)? Or something else?

Thanks.

Reply:
> Or is the concern that these values wouldn't always be accurate when the element is deeply embedded

Exactly.